File: src/eko/tools/toolList/edit-file.ts
/**
 * 文件编辑工具
 * 支持：
 * - 字面量或正则匹配
 * - 大小写可选
 * - 可选的行范围（1-based，闭区间）
 * - 预览模式（不落盘，仅返回替换统计与命中行号）
 * - 结构化 extInfo 返回详细信息
 */

import { ModifiableTool, type ToolExecutionContext } from '../modifiable-tool'
import type { ToolResult } from '@/eko-core/types'
import { writeTextFile, readTextFile } from '@tauri-apps/plugin-fs'
import { filesystemApi } from '@/api'
import { ValidationError, FileNotFoundError, ToolError } from '../tool-error'

export interface SearchAndReplaceParams {
  path: string
  // 新参数名（优先）
  search?: string
  replace?: string
  // 兼容旧参数名
  oldString?: string
  newString?: string
  // 行为控制
  useRegex?: boolean
  ignoreCase?: boolean
  startLine?: number // 1-based
  endLine?: number // 1-based，闭区间
  previewOnly?: boolean
}

/**
 * 文件编辑工具 - 支持多种编辑模式
 */
export class EditFileTool extends ModifiableTool {
  constructor() {
    super(
      'edit_file',
      `Find and replace file contents with optional regex and line range.
Behavior:
- Supports literal or regex matching, optional ignore-case
- Can restrict replacement to a specific 1-based line range [startLine, endLine]
- If previewOnly = true, will NOT write changes, only report match lines and replacement count
- If no match is found, returns a success message "No changes needed" and does not write
Notes:
- Absolute paths are required.
- Prefer using read_file to inspect context before replacing.
- Compatible with legacy parameters oldString/newString (alias of search/replace).`,
      {
        type: 'object',
        properties: {
          path: {
            type: 'string',
            description:
              'Absolute path to the file. Must be a complete path, for example: "/Users/user/project/src/main.ts", "/home/user/workspace/config.json"',
          },
          // 新参数命名（推荐）
          search: {
            type: 'string',
            description: 'Search string (literal or regex when useRegex=true).',
          },
          replace: {
            type: 'string',
            description: 'Replacement string (can be empty).',
          },
          // 兼容旧参数命名
          oldString: {
            type: 'string',
            description: 'Deprecated: same as search.',
          },
          newString: {
            type: 'string',
            description: 'Deprecated: same as replace.',
          },
          useRegex: {
            type: 'boolean',
            description: 'Enable regex matching. Default: false (literal match).',
          },
          ignoreCase: {
            type: 'boolean',
            description: 'Case-insensitive matching. Default: false.',
          },
          startLine: {
            type: 'number',
            description: 'Optional: 1-based start line (inclusive). If set, restricts replacement to this range.',
            minimum: 1,
          },
          endLine: {
            type: 'number',
            description: 'Optional: 1-based end line (inclusive). Must be >= startLine when both provided.',
            minimum: 1,
          },
          previewOnly: {
            type: 'boolean',
            description: 'If true, only preview replacement stats without writing changes.',
          },
        },
        // 仅强制要求 path，其余在运行时做更灵活的校验（兼容新旧参数名）
        required: ['path'],
      }
    )
  }

  protected async executeImpl(context: ToolExecutionContext): Promise<ToolResult> {
    const params = context.parameters as unknown as SearchAndReplaceParams

    // 基本校验与参数解析
    const path = (params.path || '').toString().trim()
    if (!path) {
      throw new ValidationError('File path cannot be empty')
    }
    if (!isAbsolutePath(path)) {
      throw new ValidationError('Path must be an absolute path')
    }

    // 兼容参数名
    const search = params.search ?? params.oldString
    const replace = params.replace ?? params.newString

    if (search === undefined) {
      throw new ValidationError('Missing required parameter: search (or oldString)')
    }
    if (replace === undefined) {
      throw new ValidationError('Missing required parameter: replace (or newString)')
    }

    const useRegex = params.useRegex === true
    const ignoreCase = params.ignoreCase === true
    const previewOnly = params.previewOnly === true

    // 行范围校验（1-based）
    const hasRange = params.startLine !== undefined || params.endLine !== undefined
    const startLine = params.startLine !== undefined ? Math.max(1, Math.floor(params.startLine)) : undefined
    const endLine = params.endLine !== undefined ? Math.max(1, Math.floor(params.endLine)) : undefined
    if (startLine !== undefined && endLine !== undefined && endLine < startLine) {
      throw new ValidationError('endLine must be greater than or equal to startLine')
    }

    // 文件校验
    const exists = await filesystemApi.exists(path)
    if (!exists) {
      throw new FileNotFoundError(path)
    }
    const isDir = await filesystemApi.isDirectory(path)
    if (isDir) {
      throw new ValidationError(`Path ${path} is a directory, cannot perform content replacement`)
    }
    if (isBinaryFile(path)) {
      throw new ValidationError(`File ${path} appears to be binary, text replacement not supported`)
    }

    // 读取原始内容
    let originalContent: string
    try {
      originalContent = await readTextFile(path)
    } catch (e) {
      throw new ToolError(`Failed to read file: ${e instanceof Error ? e.message : String(e)}`)
    }

    // 目标文本（可能是范围内的子串）
    const lines = originalContent.split('\n')
    const totalLines = lines.length

    let targetStart = 0
    let targetEnd = totalLines - 1
    if (hasRange) {
      targetStart = Math.max(0, (startLine ?? 1) - 1)
      targetEnd = Math.min(totalLines - 1, (endLine ?? totalLines) - 1)
    }

    const before = lines.slice(0, targetStart)
    const target = lines.slice(targetStart, targetEnd + 1).join('\n')
    const after = lines.slice(targetEnd + 1)

    // 构造匹配模式
    const flags = ignoreCase ? 'gi' : 'g'
    const pattern = useRegex ? new RegExp(search as string, flags) : new RegExp(escapeRegExp(search as string), flags)

    // 统计匹配（避免直接替换造成统计困难）
    const matchCount = countMatches(target, pattern)

    if (matchCount === 0) {
      const message = `No changes needed for '${path}' (0 matches).`
      return {
        content: [
          {
            type: 'text',
            text: message,
          },
        ],
        extInfo: {
          file: path,
          replacedCount: 0,
          useRegex,
          ignoreCase,
          startLine: startLine ?? null,
          endLine: endLine ?? null,
          previewOnly,
        },
      }
    }

    // 进行替换
    const modifiedTarget = (target as string).replace(pattern, replace as string)
    const modifiedContent = [...before, modifiedTarget, ...after].join('\n')

    // 预览模式：仅返回统计与命中行信息
    const matchedLines = computeMatchLineNumbers(target, pattern, hasRange ? targetStart + 1 : 1)

    // 为前端diff显示准备old和new内容
    const firstMatch = pattern.exec(target)
    const oldContent = firstMatch ? firstMatch[0] : (search as string)
    const newContent = replace as string

    if (previewOnly) {
      const message = `Preview: ${matchCount} replacement(s) will be made in '${path}'.
Lines affected (first 50): ${matchedLines.slice(0, 50).join(', ')}`
      return {
        content: [
          {
            type: 'text',
            text: message,
          },
        ],
        extInfo: {
          file: path,
          replacedCount: matchCount,
          affectedLines: matchedLines,
          useRegex,
          ignoreCase,
          startLine: startLine ?? null,
          endLine: endLine ?? null,
          previewOnly: true,
          old: oldContent,
          new: newContent,
        },
      }
    }

    // 写入文件
    try {
      await writeTextFile(path, modifiedContent)
    } catch (e) {
      throw new ToolError(`Failed to write file: ${e instanceof Error ? e.message : String(e)}`)
    }

    const message = `File edited successfully: ${path}
Status: ${matchCount} replacement(s) applied.`

    return {
      content: [
        {
          type: 'text',
          text: message,
        },
      ],
      extInfo: {
        file: path,
        replacedCount: matchCount,
        affectedLines: matchedLines,
        useRegex,
        ignoreCase,
        startLine: startLine ?? null,
        endLine: endLine ?? null,
        previewOnly: false,
        old: oldContent,
        new: newContent,
      },
    }
  }
}

export const editFileTool = new EditFileTool()

// ===== 工具函数 =====

function isAbsolutePath(p: string): boolean {
  // macOS/Linux: 以 / 开头；简单判断即可
  return p.startsWith('/')
}

function escapeRegExp(input: string): string {
  return input.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

function countMatches(text: string, regex: RegExp): number {
  if (!regex.global) {
    // 确保是全局匹配
    const flags = (regex.ignoreCase ? 'i' : '') + (regex.multiline ? 'm' : '') + (regex.dotAll ? 's' : '')
    regex = new RegExp(regex.source, 'g' + flags)
  }
  let count = 0
  let m: RegExpExecArray | null
  while ((m = regex.exec(text)) !== null) {
    count++
    if (m[0].length === 0) {
      // 防止零宽匹配死循环
      regex.lastIndex++
    }
  }
  return count
}

function computeMatchLineNumbers(text: string, regex: RegExp, baseLine: number): number[] {
  const lines: number[] = []
  if (!regex.global) {
    const flags = (regex.ignoreCase ? 'i' : '') + (regex.multiline ? 'm' : '') + (regex.dotAll ? 's' : '')
    regex = new RegExp(regex.source, 'g' + flags)
  }
  let m: RegExpExecArray | null
  while ((m = regex.exec(text)) !== null) {
    const before = text.slice(0, m.index)
    const line = baseLine + (before.split('\n').length - 1)
    lines.push(line)
    if (m[0].length === 0) regex.lastIndex++
  }
  return lines
}

function isBinaryFile(path: string): boolean {
  const ext = path.toLowerCase().split('.').pop() || ''
  const binaryExtensions = new Set([
    'jpg',
    'jpeg',
    'png',
    'gif',
    'bmp',
    'tiff',
    'tif',
    'webp',
    'ico',
    'svg',
    'mp3',
    'wav',
    'flac',
    'aac',
    'ogg',
    'm4a',
    'wma',
    'mp4',
    'avi',
    'mkv',
    'mov',
    'wmv',
    'flv',
    'webm',
    '3gp',
    'zip',
    'rar',
    '7z',
    'tar',
    'gz',
    'bz2',
    'xz',
    'exe',
    'dll',
    'so',
    'dylib',
    'app',
    'deb',
    'rpm',
    'dmg',
    'doc',
    'docx',
    'xls',
    'xlsx',
    'ppt',
    'pptx',
    'pdf',
    'ttf',
    'otf',
    'woff',
    'woff2',
    'eot',
    'db',
    'sqlite',
    'sqlite3',
    'class',
    'jar',
    'war',
    'ear',
    'pyc',
    'pyo',
    'o',
    'obj',
    'bin',
    'dat',
    'iso',
    'img',
  ])
  return binaryExtensions.has(ext)
}

File: src-tauri/src/agent/tools/builtin/edit_file.rs
use std::path::PathBuf;

use async_trait::async_trait;
use regex::RegexBuilder;
use serde::Deserialize;
use serde_json::json;
use tokio::fs;

use crate::agent::context::FileOperationRecord;
use crate::agent::persistence::FileRecordSource;
use crate::agent::state::context::TaskContext;
use crate::agent::tools::{
    RunnableTool, ToolExecutorResult, ToolPermission, ToolResult, ToolResultContent,
};

use super::file_utils::{ensure_absolute, is_probably_binary};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct EditFileArgs {
    path: String,
    search: Option<String>,
    replace: Option<String>,
    old_string: Option<String>,
    new_string: Option<String>,
    use_regex: Option<bool>,
    ignore_case: Option<bool>,
    start_line: Option<i64>,
    end_line: Option<i64>,
    preview_only: Option<bool>,
}

pub struct EditFileTool;

impl EditFileTool {
    pub fn new() -> Self {
        Self
    }
}

#[async_trait]
impl RunnableTool for EditFileTool {
    fn name(&self) -> &str {
        "edit_file"
    }

    fn description(&self) -> &str {
        "Find and replace file contents with optional regex and line range controls."
    }

    fn parameters_schema(&self) -> serde_json::Value {
        json!({
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Absolute file path"
                },
                "search": {
                    "type": "string",
                    "description": "Search string (literal or regex when useRegex=true)."
                },
                "replace": {
                    "type": "string",
                    "description": "Replacement string (can be empty)."
                },
                "oldString": {
                    "type": "string",
                    "description": "Deprecated alias of search."
                },
                "newString": {
                    "type": "string",
                    "description": "Deprecated alias of replace."
                },
                "useRegex": {
                    "type": "boolean",
                    "description": "Enable regex matching."
                },
                "ignoreCase": {
                    "type": "boolean",
                    "description": "Case-insensitive matching."
                },
                "startLine": {
                    "type": "number",
                    "minimum": 1,
                    "description": "Optional 1-based start line (inclusive)."
                },
                "endLine": {
                    "type": "number",
                    "minimum": 1,
                    "description": "Optional 1-based end line (inclusive)."
                },
                "previewOnly": {
                    "type": "boolean",
                    "description": "Preview changes without writing to disk."
                }
            },
            "required": ["path"]
        })
    }

    fn required_permissions(&self) -> Vec<ToolPermission> {
        vec![ToolPermission::FileSystem]
    }

    fn tags(&self) -> Vec<String> {
        vec!["file".into(), "edit".into(), "replace".into()]
    }

    async fn run(
        &self,
        context: &TaskContext,
        args: serde_json::Value,
    ) -> ToolExecutorResult<ToolResult> {
        let mut args: EditFileArgs = serde_json::from_value(args)?;
        let path = PathBuf::from(&args.path);

        if let Err(msg) = ensure_absolute(&path) {
            return Ok(validation_error(msg));
        }

        if is_probably_binary(&path) {
            return Ok(validation_error(format!(
                "File {} appears to be binary, text replacement not supported",
                path.display()
            )));
        }

        let metadata = match fs::metadata(&path).await {
            Ok(meta) => meta,
            Err(_) => {
                return Ok(validation_error(format!(
                    "File not found: {}",
                    path.display()
                )));
            }
        };

        if metadata.is_dir() {
            return Ok(validation_error(format!(
                "Path {} is a directory, cannot perform content replacement",
                path.display()
            )));
        }

        let search = match args.search.take().or_else(|| args.old_string.take()) {
            Some(value) => value,
            None => {
                return Ok(validation_error(
                    "Missing required parameter: search (or oldString)",
                ));
            }
        };
        let replace = match args.replace.take().or_else(|| args.new_string.take()) {
            Some(value) => value,
            None => {
                return Ok(validation_error(
                    "Missing required parameter: replace (or newString)",
                ));
            }
        };

        if let Some(start) = args.start_line {
            if start < 1 {
                return Ok(validation_error(
                    "startLine must be greater than or equal to 1",
                ));
            }
        }
        if let Some(end) = args.end_line {
            if end < 1 {
                return Ok(validation_error(
                    "endLine must be greater than or equal to 1",
                ));
            }
        }
        if let (Some(start), Some(end)) = (args.start_line, args.end_line) {
            if end < start {
                return Ok(validation_error(
                    "endLine must be greater than or equal to startLine",
                ));
            }
        }

        let original_content = match fs::read_to_string(&path).await {
            Ok(content) => content,
            Err(err) => {
                return Ok(tool_error(format!(
                    "Failed to read file {}: {}",
                    path.display(),
                    err
                )));
            }
        };

        context
            .file_tracker()
            .track_file_operation(FileOperationRecord::new(
                path.as_path(),
                FileRecordSource::ReadTool,
            ))
            .await?;

        let lines: Vec<String> = original_content
            .split('\n')
            .map(|s| s.to_string())
            .collect();
        let total_lines = lines.len();
        let last_index = total_lines.saturating_sub(1);

        let use_regex = args.use_regex.unwrap_or(false);
        let ignore_case = args.ignore_case.unwrap_or(false);
        let preview_only = args.preview_only.unwrap_or(false);

        let start_index = args
            .start_line
            .map(|v| v.max(1) as usize - 1)
            .unwrap_or(0)
            .min(last_index);
        let end_index = args
            .end_line
            .map(|v| v.max(1) as usize - 1)
            .unwrap_or(last_index)
            .min(last_index);

        let target_slice = if total_lines == 0 {
            Vec::<String>::new()
        } else {
            lines[start_index..=end_index].to_vec()
        };
        let target = target_slice.join("\n");

        let pattern = if use_regex {
            search.clone()
        } else {
            regex::escape(&search)
        };

        let regex = match RegexBuilder::new(&pattern)
            .case_insensitive(ignore_case)
            .build()
        {
            Ok(r) => r,
            Err(err) => {
                return Ok(validation_error(format!("Invalid regex: {}", err)));
            }
        };

        let match_count = regex.find_iter(&target).count();
        if match_count == 0 {
            let message = format!("No changes needed for '{}' (0 matches).", path.display());
            return Ok(ToolResult {
                content: vec![ToolResultContent::Text { text: message }],
                is_error: false,
                execution_time_ms: None,
                ext_info: Some(json!({
                    "file": path.display().to_string(),
                    "replacedCount": 0,
                    "useRegex": use_regex,
                    "ignoreCase": ignore_case,
                    "startLine": args.start_line,
                    "endLine": args.end_line,
                    "previewOnly": preview_only,
                })),
            });
        }

        let modified_target = regex.replace_all(&target, replace.as_str()).to_string();
        let modified_lines: Vec<String> =
            modified_target.split('\n').map(|s| s.to_string()).collect();

        let mut new_lines: Vec<String> = Vec::new();
        if start_index > 0 {
            new_lines.extend_from_slice(&lines[..start_index]);
        }
        new_lines.extend(modified_lines.clone());
        if total_lines > 0 && end_index + 1 < total_lines {
            new_lines.extend_from_slice(&lines[end_index + 1..]);
        }
        let new_content = new_lines.join("\n");

        let base_line = if total_lines == 0 { 1 } else { start_index + 1 };
        let mut affected_lines = Vec::new();
        for mat in regex.find_iter(&target) {
            let offset = target[..mat.start()] // safe slice, regex ensures UTF-8 boundaries
                .chars()
                .filter(|&ch| ch == '\n')
                .count();
            affected_lines.push(base_line + offset);
        }

        let old_snippet = regex
            .find(&target)
            .map(|m| m.as_str().to_string())
            .unwrap_or_else(|| search.clone());
        let new_snippet = replace.clone();

        if preview_only {
            let message = format!(
                "Preview: {} replacement(s) will be made in '{}'.\nLines affected (first 50): {}",
                match_count,
                path.display(),
                affected_lines
                    .iter()
                    .take(50)
                    .map(|line| line.to_string())
                    .collect::<Vec<_>>()
                    .join(", ")
            );
            return Ok(ToolResult {
                content: vec![ToolResultContent::Text { text: message }],
                is_error: false,
                execution_time_ms: None,
                ext_info: Some(json!({
                    "file": path.display().to_string(),
                    "replacedCount": match_count,
                    "affectedLines": affected_lines,
                    "useRegex": use_regex,
                    "ignoreCase": ignore_case,
                    "startLine": args.start_line,
                    "endLine": args.end_line,
                    "previewOnly": true,
                    "old": old_snippet,
                    "new": new_snippet,
                })),
            });
        }

        if let Err(err) = fs::write(&path, new_content).await {
            return Ok(tool_error(format!(
                "Failed to write file {}: {}",
                path.display(),
                err
            )));
        }

        context
            .file_tracker()
            .track_file_operation(FileOperationRecord::new(
                path.as_path(),
                FileRecordSource::AgentEdited,
            ))
            .await?;

        let message = format!(
            "File edited successfully: {}\nStatus: {} replacement(s) applied.",
            path.display(),
            match_count
        );

        Ok(ToolResult {
            content: vec![ToolResultContent::Text { text: message }],
            is_error: false,
            execution_time_ms: None,
            ext_info: Some(json!({
                "file": path.display().to_string(),
                "replacedCount": match_count,
                "affectedLines": affected_lines,
                "useRegex": use_regex,
                "ignoreCase": ignore_case,
                "startLine": args.start_line,
                "endLine": args.end_line,
                "previewOnly": false,
                "old": old_snippet,
                "new": new_snippet,
            })),
        })
    }
}

fn validation_error(message: impl Into<String>) -> ToolResult {
    ToolResult {
        content: vec![ToolResultContent::Error {
            message: message.into(),
            details: None,
        }],
        is_error: true,
        execution_time_ms: None,
        ext_info: None,
    }
}

fn tool_error(message: impl Into<String>) -> ToolResult {
    ToolResult {
        content: vec![ToolResultContent::Error {
            message: message.into(),
            details: None,
        }],
        is_error: true,
        execution_time_ms: None,
        ext_info: None,
    }
}
